main/java/mjc/Runner.java

```diff
@@ -0,0 +1,184 @@
+package main.java.mjc;
+
+import main.java.mjc.beans.*;
+import main.java.mjc.exceptions.CsvLineException;
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.*;
+
+public class Runner {
+    public static void main(String[] args) {
+        //OOP Task
+//        final Product product = new Product("Milk", 150);
+//        AbstractPurchase[] purchases = {
+//                new PurchaseAllDiscounted(product, 16, 10),
+//                new PurchaseAllDiscounted(product, 8, 5),
+//                new PurchaseDiscountIfGreater(product, 15, 5.825, 10),
+//                new PurchaseDiscountIfGreater(product, 10, 20, 10),
+//                new PurchaseWithExpenses(product, 5, new Euro(230)),
+//                new PurchaseWithExpenses(product, 8, new Euro(180))
+//        };
+//
+//        printPurchases(purchases);
+//        Arrays.sort(purchases);
+//        System.out.println();
+//        printPurchases(purchases);
+//        printMinimumCost(purchases);
+//
+//        PurchaseAllDiscounted searchKey = new PurchaseAllDiscounted(product, 555, 0);
+//        int index = search(purchases, searchKey);
+//        if (index < 0) {
+//            System.out.println("No purchase with cost " + searchKey.getCost() + " found.");
+//        } else {
+//            System.out.println(purchases[index]);
+//        }
+
+        //Exceptions Task
+//        Comparator<AbstractPurchase> comparator = AbstractPurchase::compareTo;
+//        PurchaseList list = new PurchaseList("main/java/mjc/resources/purchases.csv", comparator);
+//        PurchaseAllDiscounted purchase = new PurchaseAllDiscounted(product, 10, 10);
+//        list.insert(-4, purchase);
+//        list.delete(4, 7);
+//        System.out.println(list.getTotalCost());
+//        System.out.println(list.toString());
+//        list.sort();
+//        System.out.println(list.binarySearch(searchKey));;
+
+        //Collections Task
+        final String FILENAME = "main/java/mjc/resources/purchases.csv";
+        Map<AbstractPurchase, Weekday> purchasesLastDayMap;
+        Map<AbstractPurchase, Weekday> purchasesFirstDayMap;
+        Map<Weekday, PurchaseList> weekdayPurchaseEnumMap;
+        PurchaseList discountedPurchases;
+
+        try (BufferedReader reader = new BufferedReader(new FileReader(FILENAME))) {
+            purchasesLastDayMap = new HashMap<>();
+            purchasesFirstDayMap = new HashMap<>();
+            weekdayPurchaseEnumMap = new EnumMap<>(Weekday.class);
+            discountedPurchases = new PurchaseList();

```

Comment:
```suggestion
            List<PurchaseAllDiscounted> discountedPurchases = new ArrayList<>();
```
list of **the subclass** is the task requirement. 
I agree that this is artificial, but it helps to show the use of the ? metasymbol.